datafusion/src/physical_plan/expressions/distinct_expressions.rs

```diff
@@ -705,4 +844,151 @@ mod tests {
 
         Ok(())
     }
+
+    // Ordering is unpredictable when using ARRAY_AGG(DISTINCT). Thus we cannot test by simply
+    // checking for equality of output, and it is difficult to sort since ORD is not implemented
+    // for ScalarValue. Thus we check for equality via the following:
+    //   1. `expected` and `actual` have the same number of elements.
+    //   2. `expected` contains no duplicates.
+    //   3. `expected` and `actual` contain the same unique elements.
+    fn check_distinct_array_agg(
+        input: ArrayRef,
+        expected: ScalarValue,
+        datatype: DataType,
+    ) -> Result<()> {
+        let schema = Schema::new(vec![Field::new("a", datatype.clone(), false)]);
+        let batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![input])?;
+
+        let agg = Arc::new(DistinctArrayAgg::new(
+            col("a", &schema)?,
+            "bla".to_string(),
+            datatype,
+        ));
+        let actual = aggregate(&batch, agg)?;
+
+        match (expected, actual) {
+            (ScalarValue::List(Some(e), _), ScalarValue::List(Some(a), _)) => {
+                // Check that the inputs are the same length.
+                assert_eq!(e.len(), a.len());
+
+                let h1: HashSet<ScalarValue> = HashSet::from_iter(e.clone().into_iter());
+                let h2: HashSet<ScalarValue> = HashSet::from_iter(a.into_iter());
+
+                // Check that e's elements are unique.
+                assert_eq!(h1.len(), e.len());
+
+                // Check that a contains the same unique elements as e.
+                assert_eq!(h1, h2);
```

Comment:
ScalarValue does implement `PartialOrd`-- the reason it doesn't implement Ord is that it doesn't really make sense to compare different variant types -- e.g. `ScalarValue::List` and `ScalarValue::Boolean`

However, in your test cases you are always comparing the same enum, so you can do something like the following (which will panic if there is a problem comparing the `ScalarValue`s in the test):


```suggestion
                // workaround lack of Ord of ScalarValue
                let cmp = |a: &ScalarValue, b: &ScalarValue| {
                    a.partial_cmp(b).expect("Can compare ScalarValues")
                };
                e.as_mut().sort_by(cmp);
                a.as_mut().sort_by(cmp);

                // Check that the inputs are the same
                assert_eq!(e, a);
```

Reply:
Thank you! This is much nicer - I noticed the `PartialOrd` implementation but was unsure of how to actually use it.